### PR TITLE
Remove too many listeners bug

### PR DESCRIPTION
**Pull request by niamh-power**
## Problem ##

Using this library in an external project we have found out that we receive lots of events from the `onGlobalLayoutListener`
- Not all attached views receive events
- View receives more than one event

This is causing the application to trigger events for views that don't exist anymore and lingering listeners from views that never get detached

## Solution ##

- We have replaced the `onGlobalLayoutListener` by enqueing a runnable in the UI thread of the views.
- We keep track of all the listeners being attached to the `viewPager` so that the view doesn't hold invalid listeners.
- We have added a new method to the class `LandingStrip.detach(viewPager)` meant to destroy all listeners in order to prevent possible memory leaks.

Paired with @joetimmins @zegnus

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -3,7 +3,7 @@ task wrapper(type: Wrapper) {
 }
 
 allprojects {
-    version = "1.0.1"
+    version = "1.0.2"
```
Comment (blundell) [inline build.gradle]: should do this in a separate PR. i.e. this PR adds the feature, then separately we decide if we are going to up the version for a release

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java [hunk 2]
```diff
@@ -26,6 +29,7 @@
     private final ScrollOffsetCalculator scrollOffsetCalculator;
     private final FastForwarder fastForwarder;
 
+    private Set<ScrollingPageChangeListener> onScrollingPageChangeListeners;
```
Comment (blundell) [inline core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: `scrollingPageChangeListeners` saves you two chars :)

Comment (blundell) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: this set can be final, its init in the constructor

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java [hunk 3]
```diff
@@ -153,9 +155,22 @@ private void notifyDataSetChanged(PagerAdapter pagerAdapter) {
...
 
+    private void removeAllListenersFrom(ViewPager viewPager) {
+        for (ScrollingPageChangeListener listener : onScrollingPageChangeListeners) {
+            viewPager.removeOnPageChangeListener(listener);
+        }
+        onScrollingPageChangeListeners.clear();
+    }
+
+    private void trackAttachedListener(ScrollingPageChangeListener scrollingPageChangeListener) {
+        onScrollingPageChangeListeners.add(scrollingPageChangeListener);
```
Comment (blundell) [inline core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: because of the scope of this method you could simplify this var `scrollingPageChangeListener` to be just `listener`

File: core/src/main/java/com/novoda/landingstrip/TabsContainer.java [hunk 4]
```diff
@@ -47,7 +47,7 @@ boolean hasTabs() {
         return tabsContainerView.getChildCount() > 0;
     }
 
-    View getTabAt(int position) {
+    View getTabAt(final int position) {
```
Comment (joetimmins) [inline core/src/main/java/com/novoda/landingstrip/TabsContainer.java]: Does this need to be final?

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java [hunk 5]
```diff
@@ -185,8 +200,11 @@ private void forceDrawIndicatorAtPosition(int position) {
         invalidate();
     }
 
+    public void detach(ViewPager viewPager) {
```
Comment (blundell) [inline core/src/main/java/com/novoda/landingstrip/LandingStrip.java:203]: I don't see a matching attach method. what was the design choice for this?

__

How come you pass the viewpager when there is a viewpager as a field?

__

Could you use `onDetachedFromWindow` to detach, instead of allowing the client to decide when to detach?


__

maybe i should have done multiple comments :-)

Comment (zegnus) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java:203]: This is the method https://github.com/novoda/landing-strip/blob/master/core/src/main/java/com/novoda/landingstrip/LandingStrip.java#L72

```
public void attach(ViewPager viewPager) {
         attach(viewPager, viewPager.getAdapter());
     }
```

File: core/src/main/java/com/novoda/landingstrip/TabsContainer.java [hunk 6]
```diff
@@ -57,18 +57,13 @@ boolean isEmpty() {
...
-                        public void onGlobalLayout() {
-                            ViewTreeObserver observer = getTabAt(0).getViewTreeObserver();
-                            removeOnGlobalLayoutListener(observer, this);
-                            viewPager.addOnPageChangeListener(onPageChangeListener);
-
-                            onPageChangeListener.onPageScrolled(viewPager.getCurrentItem(), 0, 0);
-                        }
-                    }
-            );
+            viewPager.addOnPageChangeListener(onPageChangeListener);
```
Comment (blundell) [inline core/src/main/java/com/novoda/landingstrip/TabsContainer.java:60]: so it used to only add the listener when `onGlobalLayout` was called, now it does it instantly. Is this different?

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java (hunk 5 above)
Comment (blundell) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java:203]: I spotted the `attach` method! 👍  good to have the matching `detach` ignore Q's 1 & 3

**Review by blundell — CHANGES_REQUESTED** [review]
remove the version change before muuurge

File: core/src/main/java/com/novoda/landingstrip/TabsContainer.java (hunk 6 above)
Comment (niamh-power) [reply core/src/main/java/com/novoda/landingstrip/TabsContainer.java:60]: So `onGlobalLayout` was getting triggered multiple times (when changing tab etc too), and it was adding observers to tabs that didn't exist. Whereas now it only does it once so it shouldn't get the error we were getting.

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java (hunk 2 above)
Comment (joetimmins) [inline core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: `ScrollingPageChangeListener` doesn't have an `equals` method, so if you have two instances with the same dependencies, can they both be added? And would that be bad?

Comment (zegnus) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: Yeah they would be added, so it could happen these two things:
- More than one listener is attached when the code says that we only one one and this one is created new every time
- An old listener could receive events when that listener is not valid any more due to new data, if that's the case code might be executed in an invalid state
- We could have a potential memory leak

**Review by blundell — APPROVED** [review]

**Comment by ouchadam** [comment]
nice find! I never thought about the case where `notifydatachanged` would be called more than once

File: core/src/main/java/com/novoda/landingstrip/LandingStrip.java (hunk 2 above)
Comment (joetimmins) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: Can we make this a List then? If Set has no advantages over List, it's a little clearer to read if we use List 😇

Comment (zegnus) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: I'm not convinced on that a `list` is clearer to read than a `set`, but it's not a deal breaker in this circumstance, sure we can change that. I'm sure that if we add that `equals` we'll remember to also change the list to a set ;)

Comment (blundell) [reply core/src/main/java/com/novoda/landingstrip/LandingStrip.java]: I +1 List being clearer, there is no way a dupe can appear in this list - from the code written for `add(`, so Set confused me

**Review by joetimmins — APPROVED** [review]
